### PR TITLE
add nested tensor matmul support

### PR DESCRIPTION
**Pull request by YifanShenSZ**
There was a discussion on whether letting nested tensor `reshape` support collapsing and splitting dimension 0. The conclusion was to make reshape simple, so we need a tweaked `matmul`, which only supports 3+ dimension nonbroadcast case, i.e. a generalized `bmm`.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1122,6 +1122,10 @@
     SparseCUDA: bmm_out_sparse_cuda
     SparseCsrCUDA: bmm_out_sparse_csr_cuda
 
+- func: _NestedTensor_GeneralizedBMM(Tensor self, Tensor mat2) -> Tensor
```
Comment (cpuhrsch) [inline aten/src/ATen/native/native_functions.yaml:1125]: Let's use the dispatch system to register a kernel for matmul cc @jbschlosser

Comment (cpuhrsch) [reply aten/src/ATen/native/native_functions.yaml:1125]: I guess we can't, since matmul is composite...

Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:1125]: Right, we looked into this already - we need a custom backward and matmul is composite (it'd be cool if we could register backwards per-dispatch-key).

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -1069,6 +1220,40 @@ def test_size_dim(self):
             RuntimeError, "Given dimension 1 is irregular and does not have a size", lambda: a.size(1))
         self.assertEqual(a.size(2), 4)
 
+    def test_nested_tensor_matmul(self):
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:1223]: nit: maybe `test_nested_tensor_matmul_gradcheck` ?

File: test/test_nestedtensor.py [hunk 3]
```diff
@@ -773,29 +808,145 @@ def test_bmm(self, device, dtype):
...
+        # normal nested tensor: 4D
+        nt0 = torch.nested_tensor([torch.randn((8, 2, 4)),
+                                   torch.randn((8, 3, 7))],
+                                  device=device, dtype=dtype)
+        nt1 = torch.nested_tensor([torch.randn((8, 4, 6)),
+                                   torch.randn((8, 7, 5))],
+                                  device=device, dtype=dtype)
+        actual = torch.matmul(nt0, nt1).to_padded_tensor(0.0)
+        expect = torch.matmul(nt0.to_padded_tensor(0.0), nt1.to_padded_tensor(0.0))
+        self.assertEqual(actual, expect)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:939]: recommend adding a success case for 4D at least (unless I missed it above)

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 4]
```diff
@@ -805,6 +805,108 @@ Tensor bmm_nested(const Tensor& self, const Tensor& mat2) {
...
+  //       until we have specialized nested tensor bmm kernel
+  //       useful resource: `aten/src/ATen/native/cpu/LinearAlgebra.cpp/bmm_out_or_baddbmm_`
+  //                        `aten/src/ATen/native/cuda/Blas.cpp/baddbmm_out_cuda_impl`
+  // TODO: for 3D case, there is a warning
+  //       "An output with one or more elements was resized since it had shape [x, y],
+  //        which does not match the required output shape [1, x, y]"
+  //       because output_unbind[i].sizes() = [x, y]
+  //       but self_unbind[i] got reshaped to [1, x, z]
+  //       and mat2_unbind[i] got reshaped to [1, z, y]
+  //       after this resize got deprecated, can fix by `if 3D no reshape`
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:893]: I don't think we want to throw a warning related to internals for this case - would be confusing to a user. Can we ensure that a workaround is in place to avoid the warning?

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 5]
```diff
@@ -805,6 +805,108 @@ Tensor bmm_nested(const Tensor& self, const Tensor& mat2) {
...
+    sizemat_ptr += ndims;
+    numel += batch_size * self_size0 * mat2_size1;
+  }
+  Tensor buffer = at::empty(numel, buffer_op);
+  Tensor output = wrap_buffer(buffer, sizemat);
+  return std::make_tuple(batch_sizes, output);
+}
+}
+
+Tensor _NestedTensor_GeneralizedBMM(const Tensor& self, const Tensor& mat2) {
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:856]: suggest quick comment describing the supported shapes (>= 3 dims, same rank for both inputs)

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 6]
```diff
@@ -805,6 +805,108 @@ Tensor bmm_nested(const Tensor& self, const Tensor& mat2) {
...
+  // create a contiguous output
+  std::vector<int64_t> batch_sizes;
+  Tensor output;
+  std::tie(batch_sizes, output) = _NestedTensor_GeneralizedBMM_BatchSizes_OutputMemory(
+      self_sizes, mat2_sizes, self_buffer.options(), self_ptr->get_nested_size_tensor().options());
+  // call tensor matmul
+  // TODO: `padding nested tensor -> bmm -> remove padding` may be more efficient
+  //       until we have specialized nested tensor bmm kernel
+  //       useful resource: `aten/src/ATen/native/cpu/LinearAlgebra.cpp/bmm_out_or_baddbmm_`
+  //                        `aten/src/ATen/native/cuda/Blas.cpp/baddbmm_out_cuda_impl`
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:886]: Good TODO, I think Michael Anderson plans to upstream the fast matmul kernel

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Looks okay, some minor nits below

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 7]
```diff
@@ -1880,8 +1880,13 @@ Tensor _matmul_impl(
 
 Tensor matmul(const Tensor & tensor1, const Tensor & tensor2) {
   auto maybe_outnames = namedinference::compute_matmul_outnames(tensor1, tensor2);
-  at::Tensor unused;
-  auto result = at::native::_matmul_impl(unused, tensor1, tensor2);
+  at::Tensor result, unused;
```
Comment (jbschlosser) [inline aten/src/ATen/native/LinearAlgebra.cpp:1883]: Maybe add a note here describing why we need the is_nested() checks and possibly a TODO to remove them if we ever get per-dispatch-key backward functionality OR if we expand reshape in the future and rewrite this in terms of reshape + bmm?

File: test/test_nestedtensor.py [hunk 8]
```diff
@@ -1069,6 +1220,40 @@ def test_size_dim(self):
...
+    def test_nested_tensor_matmul_backward(self):
+        nt0 = torch.nested_tensor([torch.randn((2, 6)), torch.randn((3, 6))])
+        nt1 = torch.nested_tensor([torch.randn((6, 4)), torch.randn((6, 5))])
+        pt0 = nt0.to_padded_tensor(0.0)
+        pt1 = nt1.to_padded_tensor(0.0)
+
+        nt0.requires_grad_(True)
+        nt1.requires_grad_(True)
+        pt0.requires_grad_(True)
+        pt1.requires_grad_(True)
```
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:1247]: The reason I did this is to avoid gradient accumulation.

Concretely, if I set requires_grad in the nt factory line, then `to_padded_tensor` will be recorded in the computation graph, so the backward from pt will go all the way down to nt and ruin nt.grad

File: test/test_nestedtensor.py [hunk 9]
```diff
@@ -1069,6 +1220,40 @@ def test_size_dim(self):
...
+
+        ynt = torch.matmul(nt0, nt1)
+        ypt = torch.matmul(pt0, pt1)
+        ynt.backward(ynt.clone())
+        ypt.backward(ypt.clone())
```
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:1252]: just tried... `sum` has no nested tensor backend yet

I guess that's why Driss also uses `y.backward(y.clone())` type of testing formula

[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_nestedtensor.py:1252]: yep, right you are :(

File: test/test_nestedtensor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_nestedtensor.py:1247]: can you get around this using `with torch.no_grad()` for `the to_padded_tensor()` conversions?

File: test/test_nestedtensor.py (hunk 3 above)
Comment (jbschlosser) [reply test/test_nestedtensor.py:939]: okay, probably overkill to cover 4D components (AKA 5D nested)

File: test/test_nestedtensor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply test/test_nestedtensor.py]: yes it works

File: test/test_nestedtensor.py [hunk 10]
```diff
@@ -773,29 +808,145 @@ def test_bmm(self, device, dtype):
...
+    # cannot test torch.float16 because: RuntimeError: "bmm" not implemented for 'Half'
+    @dtypes(torch.float, torch.double)
+    def test_matmul(self, device, dtype):
+        # error case: not 3+D tensors
+        nt0 = torch.nested_tensor([], device=device, dtype=dtype)
+        nt1 = torch.nested_tensor([torch.randn(2), torch.randn(3)], device=device, dtype=dtype)
+        nt2 = torch.nested_tensor([torch.randn((2, 4)), torch.randn((3, 4))], device=device, dtype=dtype)
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"batch1 must be a 3\+D tensor",
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:842]: Do these error messages need to be updated in the tests?

**Review by jbschlosser — APPROVED** [review]
LGTM if checks pass. Thanks!

File: aten/src/ATen/native/native_functions.yaml [hunk 11]
```diff
@@ -1138,6 +1138,10 @@
...
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:1141]: This does not follow at all the naming convention form this file...

File: aten/src/ATen/native/native_functions.yaml [hunk 12]
```diff
@@ -1138,6 +1138,10 @@
...
+  dispatch:
+    NestedTensorCPU, NestedTensorCUDA: _NestedTensor_GeneralizedBMM
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:1143]: This is not needed, we can do the exact same thing as what was done for linear: https://github.com/pytorch/pytorch/pull/80397

File: tools/autograd/derivatives.yaml [hunk 13]
```diff
@@ -1337,6 +1342,10 @@
 # making it impossible (hard) to detect when it is actually a view.
 # - name: reshape(Tensor self, IntArrayRef shape)
 
+- name: _reshape_nested(Tensor self, int[] shape) -> Tensor
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1345]: Why is this here and not in the PR that adds reshape?

**Review by albanD — COMMENTED** [review]
@YifanShenSZ What is the reason for this not to follow the best practices from https://github.com/pytorch/pytorch/pull/80397 ?

**Comment by YifanShenSZ** [comment]
Thanks @albanD for clarifying the CompositeImplicit things. Will refactor `matmul` as [#80397](https://github.com/pytorch/pytorch/pull/80397): add NestedTensor key to dispatch in addition to CompositeImplicitAutograd
